### ISSUE TITLE
Scaling service according to number of active sessions

### ISSUE THREAD

**Issue by worldspawn** [kedacore/keda#1479]
### Use-Case

I have a large workload that is bound to a single session id. As a session can only be processed from a single listener (because it locks the session) scaling according to the number of messages in the queue is not... useful.

I've just added 620 messages to a queue all with the same session id and keda has scaled up the service to whopping 16 instances. But unfortunately 15 instances are doing nothing and 1 is working hard.

I'm not sure if you can interrogate the number of active sessions but if you can then...

### Specification

- [ ] Allow to specify the desired number of desired active sessions and scale based on that instead of the number of messages in a queue.

**Comment by tomkerkhove** [comment]
That's a good scenario to cover. Are you using a specific scaler or more in general?

With Azure Service Bus you'd use sessions indeed, I'm just wondering how KEDA could know when to scale and when not?

**Comment by worldspawn** [comment]
I'm using the `azure-servicebus` trigger currently. I can't see any appropriate method on the REST api that would do this but "Service Bus Explorer" has the ability to produce a list of sessions for a queue (somehow).

**Comment by worldspawn** [comment]
Seems to use this method: https://docs.microsoft.com/en-us/dotnet/api/microsoft.servicebus.messaging.subscriptionclient.getmessagesessions?view=azure-dotnet#Microsoft_ServiceBus_Messaging_SubscriptionClient_GetMessageSessions

**Comment by tomkerkhove** [comment]
Reasonable ask. So if there would be a `useSessions: true` opt-in which would scale based on metrics and sessions that would help you?

**Comment by worldspawn** [comment]
Yeah that would be great. In this particular case I wouldn't be interested in scaling based on the number of messages at all, only the number of sessions. Because of how the .net client manages concurrency I cannot imagine ever needing to scale based on both queue length and session count at the same time.

**Comment by tomkerkhove** [comment]
Makes sense, thanks for bringing this up!

**Comment by davidallyoung** [comment]
Hi everyone, just to chime in, we also have a similar use case for sessions. Some context on how we use them:

We utilize the sessions to break up workloads per tenant and help facilitate "fast lanes". It's set up in such a way that some tenants could have more sessions than others, etc. If KEDA could scale based on sessions, it would definitely simplify and make more reliable our scaling strategy.

**Comment by MCGPPeters** [comment]
Somewhat connected to this issue.... When downscaling, are receiver instances that are locked taken into account? E.g. are these instances kept alive until the session has completed and the lock has been removed?

**Comment by tomkerkhove** [comment]
No because KEDA is not aware of what instances are doing. It just tells Kubernetes to scale from x to y, and the scheduler will kill the required amount of pods.

**Comment by MCGPPeters** [comment]
That would be really nice to have, since long running sessions could theoretically span days / months ... Any way you could work around this?

**Comment by tomkerkhove** [comment]
The only way to achieve this is to change the application that's running so that when a shutdown is initiated it gracefully abandons the session for somebody else to continue later on.

**Comment by MCGPPeters** [comment]
Thanks. That is the strategy I am following atm indeed. Was hoping I could remove some code :)

**Comment by tomkerkhove** [comment]
WDYT @ahmelsayed?

**Comment by jjcollinge** [comment]
So it looks like listing sessions is partially supported in the [ASB Go SDK](https://github.com/Azure/azure-service-bus-go/blob/f9ea40a809f9b8f8f8789dc679d2088cc3935be9/message_session.go#L101) but is [limited to getting only sessions that state has been touched in the last 30 mins](https://github.com/Azure/azure-service-bus-go/issues/210). According to the JS SDK team, when they looked at implementing this a couple of years ago the AMQP operation `com.microsoft:get-message-sessions` was unreliable on ASB. So I'm not yet convinced it's possible to build this feature in a reliable way without further investment/insight from the ASB team.

**Comment by purijatin** [comment]
We have a similar use case. It would be great to have a working solution on this.

**Comment by tomkerkhove** [comment]
Are you willing to contribute this?

**Comment by raorugan** [comment]
Sure @Tom ***@***.***>, will look into this and get back on the decision.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by tomkerkhove** [comment]
Oh no need to @raorugan

**Comment by Krishnadas-KP** [comment]
@tomkerkhove Will this be implemented soon ?

**Comment by tomkerkhove** [comment]
It's looking for a contributor so if you are interested, let me know.

**Comment by b-dines** [comment]
We are also interested in this capability.  I am not sure we have the skills/bandwidth to contribute unfortunately.

**Comment by christopherlombardiconfluence** [comment]
We also have an interest in this capability.

**Comment by miqm** [comment]
@tomkerkhove Is this feature even possible to implement currently? I think that the Service Bus SDK for go doesn't provide option to get count of active sessions. There's option in C# and Java SDK (GetMessageSessions on SB client) but I don't see this in GO.

**Comment by tomkerkhove** [comment]
We have not looked in to it yet given nobody has proposed to implement it yet. If this is a known limitation, then it would be nice if you could file an issue with them!

**Comment by petemessina** [comment]
@tomkerkhove I can take a look at the change and see if I can move it forward

**Comment by aionic** [comment]
@petemessina / @tomkerkhove - I would be interested in helping as well

**Comment by tomkerkhove** [comment]
That would be wonderful! @petemessina / @aionic who wants to pick this up?

**Comment by sujesharukil** [comment]
Any movement on this feature?

**Comment by Mortana89** [comment]
Following here, we use sessions to guarantee FIFO so queue based scaling as is is indeed not useful, would love to have this!

**Comment by miqm** [comment]
@tomkerkhove - I found a ticket from last year where .NET SDK team discussed on Service Bus sessions. It seems that this feature needs to have not only SDK for GO support but also the Service Bus API: https://github.com/Azure/azure-sdk-for-net/issues/20569#issuecomment-1648266334

For scaling scenarios this feature on service bus API is kind of crucial. Do you think you could influence Service Bus team to bring this on their roadmap and provide such data?

**Comment by miqm** [comment]
Hello,

For those who are still waiting for the native support and want to try something different - I have started an open-source project with external scaler that leverages AMQP command to get count of sessions: https://github.com/miqm/session-scaler. Feel free to try it out. I will gradually try to improve it i.e. add workload identity support. All feedback (even the negative, as long as is constructive ) is appreciated!

**Comment by TOuhrouche** [comment]
Been exploring Azure ACA, we build all our multi-tenant services around sessions, which makes it a bummer as we can't use any of the queue-based scaling. That said, I was wondering if one could achieve the desired outcome by using KEDA / App Insights Metrics scaler. The services would output a "load" metric every x minutes, KEDA would scale if the load across all replicas is higher than a certain value (say 80% of max load).
